framework/include/problems/SubProblem.h

```diff
@@ -927,6 +949,30 @@ class SubProblem : public Problem
   friend class Restartable;
 };
 
+template <typename T>
+FunctorMaterialProperty<T> &
+SubProblem::declareFunctorProperty(const std::string & name, const THREAD_ID tid)
+{
+  mooseAssert(tid < _functor_material_properties.size(),
+              "Requested thread ID " << std::to_string(tid) << " too large");
+  auto & functor_material_properties = _functor_material_properties[tid];
```

Comment:
Maybe the solution to the lambda craziness could be allowing users to create their own functormaterialproperty subclasses - and this function allows them to pass in their own custom property class type - i.e.

```c++
template <typename T, typename P>
P<T> &
Subproblem::declareFunctorProperty(const std::string & name, const THREAD_ID tid)
{
  ...
  auto pr = functor_material_properties.emplace(name, std::unique_ptr<FunctorMaterialProperty<T>>(new P<T>(name));
  ...
}
```

Or does this just seem to crazy/silly?

Reply:
I think an option to sub-class is good. I really like the base `FunctorMaterialProperty` because inside of the lambda (admitting readily that the lambda syntax is not easy for users) I feel like the code looks quite a bit like current material properties. But allowing flexibility to override the behavior of the base I think is a great thing.